### DISCUSSION TITLE
Visual Scripting for Finite State Machine and Behavior Tree

### DISCUSSION THREAD

**Discussion by hyperkvlt** [godotengine/godot-proposals#7660] [Editor]
Coming from unity, I am so used  using an asset called NodeCanvas. This asset allow user to use both FSM and BT (or even creating their custom graph) for their game. And you can run both type of graph by creating subgraph using scriptableobject, this allow user to create a very complex AI, character control, and easy to understand game workflow. This asset work by running nodes inside graph, these graph are created using scriptable objects and run on scene by using graph controller, the nodes inside the graph will be run in order and changed as the condition met, and they can also use variables they create and binded using blackboard and invoke events between different graph using signal.

It would be very useful for user if godot had a solution for visual scripting of both FSM and BT to create more complex AI and work flow.

![NodeCanvas](https://github.com/godotengine/godot-proposals/assets/110368818/727f36ff-0596-4f34-842c-dff1673c826d)

**Comment by Zireael07** [comment]
There's a couple assets for this, e.g. https://github.com/imberny/VisualFSM or https://godotengine.org/asset-library/asset/817

**Reply by hyperkvlt** [reply]
Thanks! I'll check that later.

**Comment by AThousandShips** [comment]
This is best handled by a plugin or extension, Godot doesn't provide any state machines or behaviour code, and the `GraphEdit` class is exactly designed for this kind of task, in the scope of a plugin

**Reply by hyperkvlt** [reply]
You're right. But it would be great if Godot provide an official AI behavior solution such as FSM or BT

**Reply by AThousandShips** [reply]
We avoid adding too specific components, AI and behaviour code is very project specific, and there's no universal solutions

**Reply by Calinou** [reply]
See https://github.com/godotengine/godot-proposals/issues/281 and https://github.com/bitbrain/beehave.

**Reply by hyperkvlt** [reply]
> We avoid adding too specific components, AI and behaviour code is very project specific, and there's no universal solutions

I see, so the developer need to either come with their own solution or third party plugin for AI system?

**Reply by hyperkvlt** [reply]
> See #281 and https://github.com/bitbrain/beehave.

Thanks! I'll check them later

**Comment by Kchour** [comment]
Came across this project: https://github.com/limbonaut/limboai
